### PR TITLE
[cor-675]-Implemented constant folding for object splice

### PR DESCRIPTION
**Pull request by bluepal-pavani-lanka**
Improved constant folding for object literals in AQL. Object expressions containing constant object splices (...) and computed attribute names are now evaluated during query optimization when all their components are compile-time constants. This reduces unnecessary runtime object construction, resulting in more efficient query execution while preserving existing AQL semantics

### CODE REVIEW

**Comment by bluepal-avanthi-dundigala** [comment]
Overall, the implementation looks good, but the test coverage could be strengthened in a few areas. 
Consider adding tests for the new constant folding helper logic (either directly or through optimizeObject), including cases for constant objects, computed keys, constant or non-constant splices, and empty splices. It would also be good to cover the rollback path in appendObjectElementsFromConstantObject() to verify the AST is correctly restored if folding fails midway. On the JavaScript side, consider adding negative tests to ensure non-constant splices are not folded during optimization, and add tests for duplicate key behavior after constant folding (e.g: {...{a:1}, a:99} and {a:1, ...{a:2}}) to verify AQL's last-key-wins semantics are preserved. These additional tests would cover in both the optimization logic and its edge case behavior.

File: arangod/Aql/AqlValue.cpp [hunk 1]
```diff
@@ -1198,6 +1199,16 @@ AqlValue::AqlValue(int64_t low, int64_t high) {
   setType(AqlValueType::RANGE);
 }
 
+AqlValue AqlValue::fromOwnedMallocSlice(uint8_t* data, size_t length) {
+  TRI_ASSERT(data != nullptr);
+  TRI_ASSERT(length > sizeof(AqlValue));
```
Comment (adityamukho) [inline arangod/Aql/AqlValue.cpp]:
```suggestion
  TRI_ASSERT(length > sizeof(_data.inlineSliceMeta.slice));
```

Same as https://github.com/arangodb/arangodb/pull/22879/changes#diff-d55eb6ec3d368cce1bafe560b52bb5a6a53bbdb08c695fc5ee59a04d53a547a4L1266

File: arangod/Aql/Ast.cpp [hunk 2]
```diff
@@ -121,6 +122,47 @@ struct ValidateAndOptimizeContext {
...
+
+bool astNodeIsConstantObjectExpression(AstNode const* source) noexcept {
+  return source != nullptr && source->isConstant() &&
+         source->type == NODE_TYPE_OBJECT;
+}
+
+bool astNodeObjectCanFullyConstantFold(AstNode const* node) noexcept {
+  if (astNodeObjectHasObjectSplice(node)) {
+    return false;
+  }
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp]: Why are we rejecting objects containing splices?

File: arangod/Aql/Ast.cpp [hunk 3]
```diff
@@ -4154,8 +4197,192 @@ AstNode* Ast::optimizeFor(AstNode* node) {
   return node;
 }
 
+void Ast::clearObjectOptimizationFlags(AstNode* node) {
+  TRI_ASSERT(node != nullptr);
+  TRI_ASSERT(node->type == NODE_TYPE_OBJECT);
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp]: Why this assertion here? If we want to ensure the node type is only object, we should enforce that in the method contract. If that is not enforceable, we can just return with a noop.

Same for the null-check - assertions on any data should only be made within the block that created/modified that data, where the writer code can reasonably be expected to own the outcome of its own write.

What you want to do here, instead of assertion, is **validation** - in this case, trivial - just return with a noop. It should also run in production mode.

File: arangod/Aql/Ast.cpp [hunk 4]
```diff
@@ -4154,8 +4197,192 @@ AstNode* Ast::optimizeFor(AstNode* node) {
...
+    } else if (member->type == NODE_TYPE_CALCULATED_OBJECT_ELEMENT) {
+      AstNode* value = member->getMemberUnchecked(1);
+      if (value != nullptr && value->type == NODE_TYPE_OBJECT) {
+        AstNode* flattened = flattenObjectLiteralSplices(value);
+        if (flattened != value) {
+          TEMPORARILY_UNLOCK_NODE(member);
+          member->changeMember(1, flattened);
+        }
+      }
+    }
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp:4197]: There's scope for de-duplication here. The only thing apparently different in the if and else blocks are the index passed to getMemberUnchecked() and changeMember().

File: arangod/Aql/Ast.cpp [hunk 5]
```diff
@@ -4154,8 +4197,192 @@ AstNode* Ast::optimizeFor(AstNode* node) {
...
+  }
+}
+
+/// @brief flatten object literal splices into the parent object
+AstNode* Ast::flattenObjectLiteralSplices(AstNode* node) {
+  if (node == nullptr || node->type != NODE_TYPE_OBJECT) {
+    return node;
+  }
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp:4213]: This is correct validation 👍

File: arangod/Aql/Ast.cpp [hunk 6]
```diff
@@ -4154,8 +4197,192 @@ AstNode* Ast::optimizeFor(AstNode* node) {
...
+
+  size_t const n = node->numMembers();
+  containers::SmallVector<AstNode*, 8> newMembers;
+  bool changed = false;
+
+  for (size_t i = 0; i < n; ++i) {
+    AstNode* member = node->getMemberUnchecked(i);
+
+    if (member->type == NODE_TYPE_OBJECT_SPLICE) {
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp]: Can `member` ever be null? Should we guard against this possibility?

File: arangod/Aql/Ast.cpp [hunk 7]
```diff
@@ -4154,8 +4197,192 @@ AstNode* Ast::optimizeFor(AstNode* node) {
...
+      AstNode* source = member->getMemberUnchecked(0);
+      if (source == nullptr || source->type != NODE_TYPE_OBJECT) {
+        newMembers.push_back(member);
+        continue;
+      }
+
+      source = flattenObjectLiteralSplices(source);
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp:4235]: Would it be better to deep clone the input `node` before any of its members are modified? That would make the guards at:

https://github.com/arangodb/arangodb/pull/22879/changes#diff-c0cac1159dc7657a695af6abfddaecf04f4e696aee37ded0135efc839dbee64dR4228

and 

https://github.com/arangodb/arangodb/pull/22879/changes#diff-c0cac1159dc7657a695af6abfddaecf04f4e696aee37ded0135efc839dbee64dR4219

actually work as intended. Currently, they can never return true because the input node is always returned unconditionally.

File: arangod/Aql/Ast.cpp [hunk 8]
```diff
@@ -4154,8 +4197,192 @@ AstNode* Ast::optimizeFor(AstNode* node) {
...
+  }
+  return true;
+}
+
+/// @brief optimizes an object literal or an object expression
+AstNode* Ast::optimizeObject(
+    transaction::Methods& trx,
+    AqlFunctionsInternalCache& aqlFunctionsInternalCache, AstNode* node) {
+  node = flattenObjectLiteralSplices(node);
+  node = foldConstantObjectSplices(node);
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp]: Do we really need to call this after `flattenObjectLiteralSplices(node)` has already executed?

File: arangod/Aql/Ast.cpp (hunk 2 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: astNodeObjectCanFullyConstantFold() is deciding whether an object literal can be (fully constant fold) replaced entirely by a single constant VelocyPack object during optimization. It only allows objects whose members are plain object elements with constant values, and explicitly rejects:
object splices,
computed attribute names,
non-constant values

Only return true if I know I can completely fold this object using the simple object-element path.
Anything more complex returns false.

A full constant fold would need to:
evaluate the spliced object,
iterate over all its attributes,
merge them into the surrounding object,
resolve duplicate keys ("last key wins"),
preserve the exact attribute order/semantics.

File: arangod/Aql/Ast.cpp (hunk 8 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: flattenObjectLiteralSplices() does not guarantee that all NODE_TYPE_OBJECT_SPLICE nodes are eliminated. Therefore, the check in astNodeObjectCanFullyConstantFold() is not redundant.

After flattenObjectLiteralSplices() there can still be NODE_TYPE_OBJECT_SPLICE nodes whenever:
the splice source is a variable/reference,
the splice source is a bind parameter,
the splice source is a function call,
the splice source is any non-object expression,
or any case where the source is not literally an NODE_TYPE_OBJECT.

File: arangod/Aql/Ast.cpp (hunk 3 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: Replaced assertions with validating external inputs.

File: arangod/Aql/Ast.cpp (hunk 4 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp:4197]: Modified the code to eliminate the duplicate logic

File: arangod/Aql/Ast.cpp (hunk 6 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: Added null check for member.

File: arangod/Aql/Ast.cpp (hunk 7 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp:4235]: Removed the dead comparison and treating the function as an in-place mutator

File: arangod/Aql/AqlValue.cpp [hunk 9]
```diff
@@ -1198,6 +1198,16 @@ AqlValue::AqlValue(int64_t low, int64_t high) {
   setType(AqlValueType::RANGE);
 }
 
+AqlValue AqlValue::fromOwnedMallocSlice(uint8_t* data, size_t length) {
+  TRI_ASSERT(data != nullptr);
+  TRI_ASSERT(length > sizeof(_data.inlineSliceMeta.slice));
```
Comment (adityamukho) [inline arangod/Aql/AqlValue.cpp]: Please convert to validations instead of assertions.

File: arangod/Aql/Ast.cpp (hunk 2 above)
Comment (adityamukho) [reply arangod/Aql/Ast.cpp]: Why does it only reject top-level splices? Why not nested splices?

File: arangod/Aql/Ast.cpp (hunk 7 above)
Comment (adityamukho) [reply arangod/Aql/Ast.cpp:4235]: If the node is mutated in-place, why do we need to return a pointer to it?

File: arangod/Aql/Ast.cpp (hunk 8 above)
Comment (adityamukho) [reply arangod/Aql/Ast.cpp]: My question was about the call to `foldConstantObjectSplices` at line 4369, just above this comment.

File: arangod/Aql/AqlValue.cpp (hunk 9 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/AqlValue.cpp]: Replaced Assertions with validations

File: arangod/Aql/Ast.cpp (hunk 2 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: Nested splices are optimized before the outer object is constant-folded. Therefore, astNodeObjectCanFullyConstantFold() only needs to reject top-level OBJECT_SPLICE members. Nested splices are handled when their containing object is optimized.

File: arangod/Aql/Ast.cpp (hunk 8 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: Yes, foldConstantObjectSplices() is necessary. Without it, astNodeObjectCanFullyConstantFold() would reject many objects that are actually fully constant and could be folded.

flattenObjectLiteralSplices() recursively normalizes nested literal-object splices.
foldConstantObjectSplices() folds constant splices for the current object.

File: arangod/Aql/Ast.cpp (hunk 7 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp:4235]: Functionally, the return value appears redundant in the current implementation, as the function modifies the existing AstNode in place and always returns the same pointer.

From an architectural perspective, returning AstNode* is a reasonable design choice because it aligns with the optimizer framework, where some transformations modify a node in place while others may replace it with a different node.

This pattern is also common in AST optimization frameworks. Using a consistent AstNode* return type across optimization functions keeps the optimization pipeline uniform and allows callers to handle both in-place mutations and node replacements in the same way.

**Review by bluepal-pavani-lanka — COMMENTED** [review]
Went through all the review comments and updated the code accordingly

File: arangod/Aql/Ast.cpp (hunk 2 above)
Comment (adityamukho) [reply arangod/Aql/Ast.cpp]: Why not just replace the entire method body with
```
return node->isConstant() && node->isDeterministic();
```
? 

`isConstant()` already performs these same checks recursively and rejects computed keys and non-constant values (it accepts only splices whose source is a constant object literal). It also accepts a constant splice, which is safe since `Expression::execute` evaluates `OBJECT_SPLICE` correctly.

File: arangod/Aql/Ast.cpp (hunk 8 above)
Comment (adityamukho) [reply arangod/Aql/Ast.cpp]: Could you add a test case demonstrating this? Specifically, one where `foldConstantObjectSplices(node)` still changes `node` after `flattenObjectLiteralSplices(node)` has run on it?

File: arangod/Aql/Ast.cpp (hunk 2 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: We were using the astNodeObjectCanFullyConstantFold() helper in only one place, so I removed the helper and simplified the logic by using the suggested isConstant() and isDeterministic() checks directly.

Since isDeterministic() internally invokes isConstant(), and we only had a single additional condition to check before that, I simplified the logic further by avoiding the separate isDeterministic() call.

node->isConstant() &&
    (!hasFlag(DETERMINED_NONDETERMINISTIC) ||
     !hasFlag(VALUE_NONDETERMINISTIC));

File: arangod/Aql/Ast.cpp (hunk 8 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: Now I got your initial question. we can remove foldConstantObjectSplices() entirely, we have invoked only from optimizeObject() after flattenObjectLiteralSplices().
Reason:
flattenObjectLiteralSplices() already recursively removes every OBJECT_SPLICE whose source is a NODE_TYPE_OBJECT. 
foldConstantObjectSplices() only folds an OBJECT_SPLICE when its source is also a constant NODE_TYPE_OBJECT.
Hence in  foldConstantObjectSplices() successful changed = true path is not reachable.

After removing I have verified all test cases query output and also query Explain. They are working as expected.
So removed foldConstantObjectSplices() and its dependent tests from AstNodeTest.cpp.

Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: Good catch! The right questions always help us rethink things and look at them from a different perspective.

File: arangod/Aql/Ast.cpp [hunk 10]
```diff
@@ -4366,9 +4299,10 @@ AstNode* Ast::optimizeObject(
     transaction::Methods& trx,
     AqlFunctionsInternalCache& aqlFunctionsInternalCache, AstNode* node) {
   node = flattenObjectLiteralSplices(node);
-  node = foldConstantObjectSplices(node);
+  // node = foldConstantObjectSplices(node);
 
-  if (astNodeObjectCanFullyConstantFold(node)) {
+  if (node->isConstant() && (!node->hasFlag(DETERMINED_NONDETERMINISTIC) ||
+                             !node->hasFlag(VALUE_NONDETERMINISTIC))) {
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp]: I think we can drop the flag checks since isConstant() == true => node is deterministic.

File: arangod/Aql/Ast.h [hunk 11]
```diff
@@ -703,7 +710,24 @@ class Ast {
   AstNode* optimizeFor(AstNode*);
 
   /// @brief optimizes an object literal or an object expression
-  AstNode* optimizeObject(AstNode*);
+  AstNode* optimizeObject(transaction::Methods&, AqlFunctionsInternalCache&,
+                          AstNode*);
+
+  /// @brief expand constant object splices into object element nodes
+  AstNode* foldConstantObjectSplices(AstNode* node);
```
Comment (adityamukho) [inline arangod/Aql/Ast.h]: This is still declared even though its implementation has been removed.

File: arangod/Aql/Ast.h [hunk 12]
```diff
@@ -703,7 +710,24 @@ class Ast {
...
+
+  /// @brief flatten object literal splices into the parent object
+  AstNode* flattenObjectLiteralSplices(AstNode* node);
+
+  /// @brief flatten object literal member values
+  void flattenObjectLiteralMemberValues(AstNode* node);
+
+  /// @brief append object elements from a constant object expression
+  bool appendObjectElementsFromConstantObject(
+      AstNode const* source, containers::SmallVector<AstNode*, 8>& out);
```
Comment (adityamukho) [inline arangod/Aql/Ast.h]: We can remove this method decl and impl as well, since `foldConstantObjectSplices` was the only method calling it.

File: arangod/Aql/Ast.cpp [hunk 13]
```diff
@@ -121,6 +122,21 @@ struct ValidateAndOptimizeContext {
...
+      return true;
+    }
+  }
+  return false;
+}
+
+bool astNodeIsConstantObjectExpression(AstNode const* source) noexcept {
+  return source != nullptr && source->isConstant() &&
+         source->type == NODE_TYPE_OBJECT;
+}
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp]: Please remove this method.

File: tests/Aql/AstNodeTest.cpp [hunk 14]
```diff
@@ -115,6 +136,530 @@ AstNode const* expectObjectObject(AstNode const* object,
...
+      parseAndOptimizeReturnExpression(_server, "RETURN { ...{}, x: 1 }");
+  ASSERT_NE(nullptr, objectNode);
+  ASSERT_EQ(NODE_TYPE_OBJECT, objectNode->type);
+  EXPECT_TRUE(objectNode->isConstant());
+  EXPECT_FALSE(objectHasObjectSplice(objectNode));
+  ASSERT_EQ(1, objectNode->numMembers());
+  expectObjectInt(objectNode, "x", 1);
+}
+
+TEST_F(AstNodeTest, appendObjectElementsFromConstantObjectRollsBackOnFailure) {
```
Comment (adityamukho) [inline tests/Aql/AstNodeTest.cpp]: Please delete this test.

File: tests/Aql/AstNodeTest.cpp [hunk 15]
```diff
@@ -115,6 +136,530 @@ AstNode const* expectObjectObject(AstNode const* object,
...
+    }
+  }
+  return false;
+}
+
+/// @brief Build an object that still reports isConstant() because the constant
+/// flags were cached before a calculated key was appended. This models a
+/// mid-fold failure inside appendObjectElementsFromConstantObject(): some
+/// OBJECT_ELEMENT members are appended successfully, then folding fails.
+AstNode* makePoisonedConstantObject(Ast* ast) {
```
Comment (adityamukho) [inline tests/Aql/AstNodeTest.cpp]: Please delete this method.

File: arangod/Aql/Ast.cpp [hunk 16]
```diff
@@ -4159,8 +4176,144 @@ AstNode* Ast::optimizeFor(AstNode* node) {
...
+  return true;
+}
+
 /// @brief optimizes an object literal or an object expression
-AstNode* Ast::optimizeObject(AstNode* node) {
+AstNode* Ast::optimizeObject(
+    transaction::Methods& trx,
+    AqlFunctionsInternalCache& aqlFunctionsInternalCache, AstNode* node) {
+  node = flattenObjectLiteralSplices(node);
+  // node = foldConstantObjectSplices(node);
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp]: Please delete this comment.

File: arangod/Aql/Ast.cpp [hunk 17]
```diff
@@ -4159,8 +4176,144 @@ AstNode* Ast::optimizeFor(AstNode* node) {
...
+    AstNode* value = member->getMemberUnchecked(valueIndex);
+    if (value == nullptr || value->type != NODE_TYPE_OBJECT) {
+      continue;
+    }
+
+    flattenObjectLiteralSplices(value);
+  }
+}
+
+/// @brief flatten object literal splices into the parent object
```
Comment (adityamukho) [inline arangod/Aql/Ast.cpp:4208]: Please add a comment saying it currently always mutates in place and returns the same node.

File: arangod/Aql/Ast.h (hunk 12 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.h]: Yes, Removed this unused helper.

File: arangod/Aql/Ast.h (hunk 11 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.h]: Removed this unused helper.

File: arangod/Aql/Ast.cpp (hunk 13 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: Yes, Removed this unused helper.

File: arangod/Aql/Ast.cpp (hunk 16 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: Yes, Removed this commented code.

File: arangod/Aql/Ast.cpp (hunk 10 above)
Comment (bluepal-pavani-lanka) [reply arangod/Aql/Ast.cpp]: Got it, Removed redundant determinism flag checks

**Review by adityamukho — APPROVED** [review]
